[GSoC] Added a preliminary Lark LaTeX parser implementation.

wermos opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Resumes the work of https://github.com/sympy/sympy/pull/19825.

As requested by @asmeurer and @Upabjojr, I have made a draft PR on which all my work will be pushed. I have marked it as a draft because it is still a WIP.

#### Brief description of what is fixed or changed

This is the v1 rewrite of the current ANTLR-based LaTeX parser. It contains most of the features that the ANTLR-based parser supports. Notable exceptions are:
* Interpreting `\frac{d}{dx}` as a derivative.
* Interpreting integrals with the `d` operator in a fraction, like `\int \frac{dz}{z}`.
* Support for Bra and Ket notation.
* Recognizing variables like `y_1''` and interpreting them as symbols correctly
* Recognizing expressions like `f'_1(x)` as function applications.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* parsing
  * Added a Lark-based LaTeX parser.
<!-- END RELEASE NOTES -->

wermos committed:
Initial LaTeX parser commit.

wermos committed:
Refactored the code a bit.

wermos committed:
Added a bunch of tokens to the `latex.lark` file.

wermos committed:
Refactored the Lark-based parser so that Lark is imported only once.

Instead of being imported every time the parsing function is called, it will be imported just once when the file is run for the first time.

wermos committed:
Expanded the grammar a bit.

wermos committed:
Added comments and separated the symbols into distinct groups. Also added a couple more left and right symbols, and expanded the grammar a little.

wermos committed:
Enabled debug level logging.

wermos commented:
@sylee957 As you suggested in [the previous PR](https://github.com/sympy/sympy/pull/25219#issuecomment-1620942627), I've enabled the logger in this branch.

Upabjojr commented on sympy/parsing/latex/lark/latex_parser.py:
```diff
@@ -0,0 +1,43 @@
+                        maybe_placeholders=False,
+                        keep_all_tokens=True)
+
+    print("expression =", s)
```
these `print` statements should not be merged.

wermos replied:
As these are temporary, I'm marking this as resolved. They will be removed at the end.

Upabjojr commented on sympy/parsing/latex/lark/latex_parser.py:
```diff
@@ -0,0 +1,43 @@
+
+
+if __name__ == "__main__":
+    parse_latex_lark(r"a \;\thickspace * b")
```
this should be removed

wermos replied:
Resolved during the weekly meeting.

Upabjojr commented on sympy/parsing/tests/test_latex_lark.py:
```diff
@@ -0,0 +1,272 @@
+lark = import_module("lark")
+
+if not lark:
+    disabled = True
```
is this variable needed? Correct way to use it would be `disabled = lark is None`

wermos replied:
I'm not sure if it is needed or not. This is how ANTLR tests are disabled if the `antlr4-python3-runtime` is not found, in the current [master](https://github.com/sympy/sympy/blob/46558492a74bf2a324f9560af008b6c3ee587a6c/sympy/parsing/tests/test_latex.py#L27-L29).

Since we want to do the same thing if Lark is not found, I added it.

wermos replied:
This is checked for in [this function](https://github.com/sympy/sympy/blob/0ec134d277f14961e8b6629182895fc986658c2e/sympy/conftest.py#L67-L74).

wermos committed:
Added support for signed integers and floats, fixed the issue with `L_BAR`, and added to the grammar a little.

Upabjojr committed:
rewriting during team meeting

Upabjojr commented:
@wermos any updates?

wermos committed:
Fixed some formatting and eliminated the source of ambiguous parse trees.

wermos committed:
Converted the logger into an optional keyword-only argument.

wermos commented on sympy/parsing/latex/lark/latex.lark:
```diff
@@ -0,0 +1,175 @@
+FUNC_COT:  "\\cot"
+
+// inverse trigonometric functions
+FUNC_ARCSIN: "\\arcsin" | "\\sin^{-1}"
```
Good idea. I'll implement it like that.

wermos replied:
I still need to handle things like `\sin^2 x`. This is correctly handled in the ANTLR parser (I checked), but there are no tests for ensuring this behavior in [`test_latex.py`](https://github.com/sympy/sympy/blob/46558492a74bf2a324f9560af008b6c3ee587a6c/sympy/parsing/tests/test_latex.py) so it is not a priority. My first priority is passing the test cases in the `test_latex.py`.

wermos replied:
I've added this feature.

wermos committed:
Changed the name of the start string and added a temporary sanity test.

wermos committed:
Added more content to the grammar.

Removed the unnecessary `CMD_DBINOM` and `CMD_TBINOM` symbols. Added the `CMD_MATHIT` symbol.

Removed some unnecessary comments.

Refactored the structure of the grammar, and also added support for absolute value, ceiling, floor functions.

Added support for relations containing expressions on either side.

Added support for multi-letter symbols.

Added support for fractions and binomial expressions (like "\binom{n}{k}").

Added scaffolding for trigonometric expressions, and support for the sine function.

Upabjojr commented on sympy/parsing/latex/lark/latex.lark:
```diff
@@ -133,36 +135,67 @@ group: L_PAREN expression R_PAREN
+    | general_binomial
 
-binomial: CMD_BINOM L_BRACE group R_BRACE L_BRACE group R_BRACE
+general_binomial.0: CMD_BINOM group_curly_parentheses group_curly_parentheses expression?
```
We should create two kinds of expressions: the ones that have higher priority than `\binom` and the ones that have lower. That was the idea of `expression_core`... `expression_core` should contain all expressions that have higher priority than the backslash operators.

Generic `expression` should be either an `expression_core` or any kind of expressions that have lower priority than backslash operators.

This way should work.

wermos replied:
Using a named rule `A` in a different rule `B` does not mean that `A` has a higher priority than `B`. That is to say, `expression_core` does not have a higher precedence than `expression`. Both have equal precedence, which is the root reason for many of the ambiguous parses.

Assigning `expression_core` a higher precedence than `expression` _is_ possible, but there are two things to keep in mind:
1. Assigning priorities to rules is only possible using the Earley parser and not the LALR parser. Since you mentioned that you think that moving to the LALR parser might be a better option, I am mentioning this.
2. If we go with the rule priorities route, then we lose the ability to inform the user that their input expression is ambiguous. Using rule priorities requires to use `ambiguity=resolve`, which means that any ambiguity in the parsing is resolved automatically (using rule priorities, if that is available). If we keep `ambiguity=explicit`, we retain the ability to inform the user of ambiguous parses, but lose the ability to use rules with different priorities.

wermos replied:
I'm also marking this as resolved, because we discussed how to solve this problem during the team meeting, and the way to solve this is to use the same strategy but for binomial expressions.

wermos committed:
Added more content to the grammar.

Removed the ``"\\XXX^{-1}"` string from the inverse trigonometric function symbols.

Modified the `relation` terminal to take any arbitrary expression on either side of the relation. This allows for things like `"a^2 + b^2 = c^2"`.

Broke out the implicit multiplication rules into their own separate rule, because of the number of special cases it has.

Added more cases to the implicit multiplication rule.

Added support for unary minus (under the `unary_neg` rule), and for superscripts.

Added support for inverse trigonometric functions, hyperbolic trigonometric functions, inverse hyperbolic trigonometric functions, and exponential functions in the `function` rule. Added rules for all the functions which fall under those categories (like cos, tan, etc.).

wermos committed:
Added more content to the grammar.

Added support for logarithm expressions, square root expressions, and factorials.

wermos committed:
Added more content to the grammar.

Added support for most integral types, and for complex conjugation.

wermos committed:
Added more content to the grammar.

Added support for sums and products.

wermos committed:
Added a function (temporarily) to easily check the number of passing test cases.

wermos committed:
Added a couple of LaTeX commands to the list of ignored tokens and deleted some comments.

The commands ``"\nolimit"` and `"\displaystyle"` are used to tell LaTeX how to display the equation. Hence, ignoring them is fine because they only convey information about displaying equations, and not semantic information.

wermos committed:
Removed a bunch of unnecessary comments.

wermos committed:
Started writing the transformer class.

wermos committed:
Started writing the transformer class.

Upabjojr commented:
Do we have some reference containing the operator precedence in LaTeX?

wermos committed:
Added code to show more descriptive output so that we can see what the state of the Transformer is.

wermos committed:
Added code to show more descriptive output so that we can see what the state of the Transformer is.

wermos committed:
Added more functions to the transformer class.

wermos committed:
Added more functions to the transformer class.

wermos commented on sympy/parsing/latex/lark/latex.lark:
```diff
@@ -0,0 +1,270 @@
+
+integral: FUNC_INT expression "d"oneletter_symbol
+    | FUNC_INT (UNDERSCORE (group_curly_parentheses | atomic_expr) CARET (group_curly_parentheses | atomic_expr))? expression? "d"oneletter_symbol
+    | FUNC_INT (CARET (group_curly_parentheses | atomic_expr) UNDERSCORE (group_curly_parentheses | atomic_expr))? expression? "d"oneletter_symbol
```
`expression_core` also allows for `group_round_parentheses` which is not a valid option in this scenario.

wermos replied:
As you mentioned [here](https://matrix.to/#/!ieKAyoLMgajmgkdQZE:matrix.org/$sSGfnbR9sCvD06rPCApkMbi4aMeKgp2ok7CtjgRR1AQ?via=matrix.org&via=gitter.im), I removed `group_round_parentheses` from the definition of `expression_core` and made this change.

Upabjojr commented on sympy/parsing/latex/lark/latex.lark:
```diff
@@ -0,0 +1,270 @@
+
+fraction: basic_fraction
+    | simple_fraction
+    | general_fraction
```
what about just having:
```
fraction: CMD_FRAC expression_core expression_core
```
that is, `expression_core` should contain all elements that have higher priority than `\frac`.

wermos replied:
Setting aside the `group_round_parentheses` discussion, that would still not work. If we make this change, then in `\frac12`, the `12` would parse as `number`, when we need it to parse as `DIGIT DIGIT` in this case.

A similar thing would happen for `\frac1234` whereas we need the `1234` to parse as `DIGIT DIGIT expression` where the expression contains `34` as a `number`.

wermos commented:
> Do we have some reference containing the operator precedence in LaTeX?

I'm not exactly sure what you mean, but you can use [this](https://www.overleaf.com/learn/latex/Operators) as a reference for operators in LaTeX.

wermos committed:
Modified the definition of `expression_core`.

Upabjojr committed:
Team meeting: added way to handle operator priority and left/right associativity, applied functions

Upabjojr committed:
added support for limits

wermos committed:
Refactored the tests into separate categories.

wermos committed:
Added a temporary helper Python file.

The file gives us helpful output after running all the existing tests, and also has a `tricky_tests` function which allows us to stress test the WIP implementation using tests that we have usually struggled with.

wermos committed:
Minor fixes.

Changed the name of the `tricky_tests` function to have `test` in the front.

Fixed PEP8 issues.

wermos committed:
Fixed PEP8 issues and removed a now-unnecessary function.

wermos committed:
Removed an unnecessary comment and improved the name of a rule.

wermos committed:
Added support for Greek symbols.

wermos commented on sympy/parsing/latex/lark/latex.lark:
```diff
@@ -0,0 +1,323 @@
+MUL_SYMBOL: MUL | CMD_TIMES | CMD_CDOT
+
+// Greek symbols
+// TODO: Shouold we include the uppercase variants for the symbols where the uppercase variant doesn't have a separate meaning?
```
> // TODO: Shouold [sic] we include the uppercase variants for the symbols where the uppercase variant doesn't have a separate meaning

@Upabjojr could you comment on this?

wermos replied:
I'll resolve all the Greek symbol related queries, and move them into an issue of their own. They are low-priority, and won't require much work to add. All I really need is clarification on what features we want and which ones we don't want.

The Greek symbol stuff isn't important to the v1 Lark-based parser PR.

wermos commented on sympy/parsing/latex/lark/latex.lark:
```diff
@@ -0,0 +1,323 @@
+ALPHA: "\\alpha"
+BETA: "\\beta"
+GAMMA: "\\gamma"
+DELTA: "\\delta" // TODO: Should this be included? Delta usually denotes other things.
```
> // TODO: Should this be included? Delta usually denotes other things.
@Upabjojr what are your thoughts on this?

wermos commented on sympy/parsing/latex/lark/latex.lark:
```diff
@@ -0,0 +1,323 @@
+ZETA: "\\zeta"
+ETA: "\\eta"
+THETA: "\\theta" | "\\vartheta"
+// TODO: Should I add iota to the list?
```
@Upabjojr What about this? Should `\iota` be added? It's a valid LaTeX command, according to [this reference](https://www.overleaf.com/learn/latex/List_of_Greek_letters_and_math_symbols), but I've rarely seen it used as a bona fide variable. At best, I've seen people use it to denote the imaginary number `i`.

wermos commented on sympy/parsing/latex/lark/latex.lark:
```diff
@@ -0,0 +1,323 @@
+THETA: "\\theta" | "\\vartheta"
+// TODO: Should I add iota to the list?
+KAPPA: "\\kappa"
+LAMBDA: "\\lambda" // TODO: What about the uppercase variant?
```
@Upabjojr What do you think of this? I think it's fine but I just wanted to cross-check with you.

wermos commented on sympy/parsing/latex/lark/latex.lark:
```diff
@@ -0,0 +1,323 @@
+MU: "\\mu"
+NU: "\\nu"
+XI: "\\xi"
+// TODO: Should there be a separate note for transforming \pi into sympy.pi?
```
@Upabjojr For this one, I think doing it like this is a good idea. What do you think? I think that'll be a valuable feature as there is already an open issue for handling it specially: #20306.

wermos commented on sympy/parsing/latex/lark/latex.lark:
```diff
@@ -0,0 +1,323 @@
+XI: "\\xi"
+// TODO: Should there be a separate note for transforming \pi into sympy.pi?
+RHO: "\\rho" | "\\varrho"
+// TODO: What should we do about sigma?
```
@Upabjojr I'm not sure whether we should have lowercase sigma or not. It's usually used to denote standard deviation, but I'm not sure how relevant that is here.

wermos committed:
Added a test.

Upabjojr committed:
pretty_print_lark_trees

Upabjojr committed:
Team meeting: fixing some ambiguities; better pretty lark printer

wermos committed:
Started working on the Transformer.

Folded a bunch of nodes in the Lark grammar to simplify the Transformer class.

Revamped the `temporary_helper.py` file to emit more useful stats while working on the transformer.

Updated the old transformer code I had written (for experimenting purposes) for the new grammar, and removed a bunch of code for the now-useless nodes.

wermos committed:
Merge remote-tracking branch 'origin/latex-parser-transformer' into latex-parser-transformer

# Conflicts:
#	sympy/parsing/latex/lark/latex_parser.py
#	sympy/parsing/tests/test_latex_lark.py

wermos committed:
Spaced out the pretty_print_lark_tree function and slightly increased readability.

wermos committed:
Added support for trigonometric and hyperbolic trigonometric expressions in the transformer.

wermos committed:
Added support for a few more functions and changed the argument for each member function from `token` to `tokens`.

wermos committed:
Fixed the code for the `conjugate` node and added the `square_root` node.

wermos committed:
Added the absolute value and exponential nodes.

wermos committed:
Added more features to the transformer.

Improved a comment in the DummyTransformer class.

Added the code for the `GREEK_SYMBOL` node.

Added the code for the `multiletter_symbol` node.

Added Transformer support for the `div` node.

Added Transformer support for the `log` node, completely handling the different cases.

wermos committed:
Added a bunch of transformer node stubs.

wermos committed:
Added transformer support for the `relation` node.

wermos committed:
Changed the stub code from `pass` to `print(tokens)` so that I can see the token format and write the transformer code according to the format of the list of parameters the function receives.

wermos committed:
Added transformer code for the `superscript` and `sub` nodes.

wermos committed:
Wrote part of the code required to support function application.

wermos committed:
Added support for integrals in the transformer.

wermos committed:
Moved the definition of the `list_of_expressions` node to be closer to the `function_applied` node.

Also removed the definition of the `comma_operator` entirely, as it would lead to extra code in the transformer for no benefit.

wermos committed:
Added support for the function application node.

wermos committed:
Slightly modified the transformer for the logarithm node.

wermos committed:
Modified the `group_parentheses_special` to make it easier to write the transformer.

wermos committed:
Added `\leqslant` and `\geqslant` to the definition of the less-than-or-equal-to and greater-than-or-equal-to operators.

wermos committed:
Folded the `implicit_multiplication` node to simplify the transformer code.

wermos committed:
WIP Commit of the `summation` and `product` node rewrite.

wermos committed:
Added support for implicit multiplication, partially wrote the transformer code for limits, rewrote the transformer code for `list_of_expressions`.

wermos committed:
Added a couple of aliases for `CMD_FRAC` and `FUNC_INT`.

These were requested for in https://github.com/sympy/sympy/issues/25364#issuecomment-1656586985.

wermos committed:
Updated the grammar.

Removed a bunch of unnecessary imports, modified the definition of the `number` node, and removed a bunch of unnecessary comments.

Modified the definition of `sub` to have two specific cases.

Folded `fraction` so that it doesn't show up in the parse tree.

Broke up the definition of `fraction` and `binomial` into 3 separate cases.

wermos committed:
Updated the grammar.

Added support for `\text{d}` and `\mathrm{d}` in integral expressions.

wermos committed:
Updated the grammar.

Fixed the issues with `summation` and `product` by correcting their precedence level.

wermos committed:
Added more features to the transformer.

Added support for the `DIGIT` node.

Modified and finalized the transformer code for the `number` node.

Added support for fractions.

Updated a comment in the `integral` node.

Added support for the `group_curly_parentheses_special` node, which is required for the
`summation` and `product` nodes.

Updated a comment in the `remove_tokens` function.

Added a couple of short forms for commonly-used debug and testing functions.

wermos committed:
Simplified the grammar and transformer a little.

wermos committed:
Added support for the `binomial` node.

wermos committed:
Refactored the grammar a bit, by moving some definitions around.

wermos committed:
Expanded the definitions of `_basic_fraction`, `_simple_fraction`, `_basic_binomial`, and `_simple_binomial` a little.

wermos committed:
Updated the fraction-related tests.

wermos committed:
WIP commit for variable subscript bug fix.

wermos committed:
Updated the `square_root` node to use the `root` function.

wermos committed:
Moved some misplaced tests, and added some handling for some failing tests since we will deal with those features later.

wermos committed:
Fixed some bugs in the `integral` code for the transformer.

Removed a ill-formed integral-related test, and updated some integral-related tests.

Updated a sum-related test.

Added some expected failures for the integral-related tests.

Added the `XFAIL` marker to the derivative tests since that feature is yet to be added.

Added the `XFAIL` marker to the miscellaneous tests since those expressions are suspect.

Added the `XFAIL` marker to the spacing-command-related tests since there is an unsolved bug there.

wermos committed:
Marked the applied function tests as `XFAIL` due to a mysterious bug, and updated  the logarithm tests in `COMMON_FUNCTION_EXPRESSION_PAIRS`.

wermos committed:
Fixed the issue with symbol subscripts, and made most of the symbol-related tests pass.

wermos committed:
Fixed the issue with multi-letter symbols, and got the parser to pass a test that was previously failing.

Also committed some WIP work on the `limit` node.

wermos committed:
Added support for the `limit` node to the transformer.

Modified the definition for the limit node in the grammar slightly.

Updated one of the limit-related tests.

wermos committed:
Added support for powers of trigonometric functions.

wermos committed:
Fixed code quality and removed some testing scaffolding.

wermos committed:
Removed the temporary_helper file from source control.

wermos committed:
Fixed some flake8 issues.

wermos commented:
@Upabjojr @sylee957 I think this PR is now ready for code review. Feedback would be appreciated.

wermos committed:
Fixed the "could not find parse_latex_lark" issue.

wermos committed:
Added lark to the list of optional dependencies to download.

wermos committed:
Made the Lark parser tests run only if Lark is present.

wermos committed:
Merge branch 'master' into latex-parser-from-scratch

wermos committed:
Added a missing imoprt.

wermos commented:
For some reason it keeps trying to run the lark tests in `tests-latest` even though it should only be tested with all the optional dependencies. I'm not sure why.

Upabjojr committed:
fix for mypy

Upabjojr committed:
fixing mypy

Upabjojr committed:
added missing import

wermos committed:
Added the `.lark` files to `setup.py`.

wermos committed:
Removed some unnecessary code.

wermos committed:
Fixed the path to the `.lark` files in `setup.py`.

Upabjojr committed:
Restructuring Lark LaTeX parser code

wermos committed:
Removed some debug print stuff.

wermos committed:
Fixed the library name in the comments.

wermos committed:
Improved the consistency of the quotation marks used.

wermos committed:
Improved a comment.

wermos committed:
Minor improvements.

wermos committed:
Fixed code quality.

wermos committed:
Modified a comment.
